File: 01-arrays_&_hashing/00001-Two_Sum/00001-two_sum.py
# 1: Two Sum
# https://leetcode.com/problems/two-sum/

class Solution:
    # SOLUTION
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        map = {}
        for i, n in enumerate(nums):
            m = target - n
            if m in map:
                return [map[m], i]
            else:
                map[n] = i


if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [3,2,4]
    target: int = 6

    # OUTPUT
    result = o.twoSum(nums, target)
    print(result)

File: 01-arrays_&_hashing/00001-Two_Sum/00001-two_sum.cpp
// 1: Two Sum
// https://leetcode.com/problems/two-sum/

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    // SOLUTION
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> map;

        for (int i=0;;i++) {
            auto it = map.find(target - nums[i]);

            if (it != map.end())
                return {it->second, i};
            
            map[nums[i]] = i;
        }
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {3,2,4};
    int target = 6;

    // OUTPUT
    auto result = o.twoSum(nums, target);
    cout<<"["<<result[0]<<", "<<result[1]<<"]"<<endl;

    return 0;
}
